framework/src/utils/PetscSupport.C

```diff
@@ -405,7 +405,7 @@ petscNonlinearConverged(SNES snes,
   // Whether or not to force SNESSolve() take at least one iteration regardless of the initial
   // residual norm
   PetscBool force_iteration = PETSC_FALSE;
-#if !PETSC_RELEASE_LESS_THAN(3, 8, 4)
+#if !PETSC_VERSION_LESS_THAN(3, 8, 4)
```

Comment:
Ask @permcody  for suggestions.

Reply:
I yield to you or @jwpeterson to make these decisions. If you think this is correct then by all means, merge this change.

Reply:
As I understand it, we cannot switch `PETSC_VERSION` macros to `PETSC_RELEASE` macros until the code in question makes it into a release. If @fdkong says this is in 3.8.4 then it works for me.

Reply:
I do not know why `PETSC_RELEASE_LESS_THAN` does not work with my particular version of PETSc. But looks to me almost everywhere else we are using `PETSC_VERSION_LESS_THAN` and forcing users to check out a particular version of PETSc to make MOOSE work sounds not right. I guess I honestly do not know the difference between `PETSC_RELEASE_LESS_THAN` and `PETSC_VERSION_LESS_THAN`.

Reply:
@jwpeterson thanks for the explanation! I assume a hash in between two releases will have the version number equal to the version number of the first release. Since RELEASE should be used for advanced features, 3.8.4 and 3.5.0 seems way behind current petsc release and thus need to be fixed. I will update to master.

Reply:
> Since RELEASE should be used for advanced features, 3.8.4 and 3.5.0

Yeah, as features appear in released versions of PETSc, we should update these macros.

> I will update to master.

Sounds good, we don't really support using older versions of PETSc master, and things can break (as you have seen) if you get behind.

Reply:
Thanks. I am thinking `#if !PETSC_RELEASE_LESS_THAN(3, 8, 4)` may not be right, what we want is `#if !PETSC_VERSION_LESS_THAN(3, 8, 4) && !PETSC_VERSION_RELEASE`.  `!PETSC_VERSION_RELEASE` does not mean current master as @fdkong is thinking. I would avoid using this to avoid brain damage...

Reply:
> Thanks. I am thinking #if !PETSC_RELEASE_LESS_THAN(3, 8, 4) may not be right, what we want is #if !PETSC_VERSION_LESS_THAN(3, 8, 4) && !PETSC_VERSION_RELEASE. 

Nope. we want '#if !PETSC_VERSION_LESS_THAN(3, 8, 4) || !PETSC_VERSION_RELEASE.' Literally, we want PETSc-3.8.4 or higher or the current PETSc-master. The current PETSc-master may still have a tag like 3.8.3 or 3.8.2.   

> !PETSC_VERSION_RELEASE does not mean current master as @fdkong is thinking.

There is no way to check if the PETSc-master is current or not (changing everyday but tag may stay  the same). Just no way to do that.  

> I would avoid using this to avoid brain damage...

For developers only or advanced users. I agree normal users may get stuck here. But we do not encourage normal users to use PETSc-master.

Reply:
> Literally, we want PETSc-3.8.4 or higher or the current PETSc-master.

They key word here is "current". I think what @YaqiWang proposes is better, since it at least restricts us to versions of PETSc master newer than a specific release.

Reply:
> I think what happened is we wrote the PETSC_RELEASE_LESS_THAN thing not realizing that people would use its negation, which doesn't really lead to a useful check.

We realized this, but could not come up a useful check. At least I have no idea how to check if PETSc-master is up to date.  Check commit date? or time stamp?

Reply:
At least we want higher than PETSc-3.8.4 AND not released, which include master. Since we cannot check CURRENT, we should not use OR. `!PETSC_VERSION_RELEASE` is not a valid check.

Reply:
I cannot think of a good solution. I suggest we stick to `PETSC_VERSION_LESS_THAN` in MOOSE until a new release including the capability is coming in PETSc. We definitely can have a PR testing the new capability which will be tested against the current master of petsc.